### PR TITLE
Happiness v2

### PR DESCRIPTION
**Pull request by birdsarah**
This switches from using periodic callback to update the plots to having django update the plots when there is new data.

In Happiness v1, the database was being pinged every 2 seconds. This new technique should be far more efficient and scalable and is a net reduction in code (you can't quite see this from the PR because other artefacts have been added).

There is a notebook that has some narrative about things that have been learned making v1 and v2 and things that could still be improved.

### CODE REVIEW

**Comment by bryevdv** [comment]
Some small suggestions (maybe other PR):

* would be nice if the data was always the "last X days, regardless of when the user runs the demo
* didn't realize date did not get incremented when happiness logged, was confusing, suggest auto-incrementing date
* does not seems to be auto-ranging when new points are added?
* reset seems to weirdly shrink plot size (!)

**Comment by bryevdv** [comment]
Things where it would just be good to know more:

* is a little slow to log happiness. Where is the slowness occurring?
* lots of closed connections logged, would just like to know what's going on

File: happiness/happiness/bokeh_utils.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+def update_bokeh_sessions(user_sessions):
+    for us in user_sessions:
+        with closing(pull_session(session_id=us.bokeh_session_id)) as session:
+            if len(session.document.roots) == 0:
+                # In this case, the session_id was from a dead session and
+                # calling pull_session caused a new empty session to be
+                # created. So we just delete the UserSession and move on.
+                # It would be nice if there was a more efficient way - where I
+                # could just ask bokeh if session x is a session.
+                us.delete()
```
Comment (bryevdv) [inline happiness/happiness/bokeh_utils.py]: Just to be clear, this won't work with multiple servers behind a load balancer / proxy, right?

Comment (birdsarah) [inline happiness/happiness/bokeh_utils.py]: what? multiple django servers or multiple bokeh servers? `us` is a `UserSession` django model instance.

**Comment by birdsarah** [comment]
Thanks for the look over @bryevdv - some follow-up questions

>Things where it would just be good to know more:
>
>    is a little slow to log happiness. Where is the slowness occurring?
>    lots of closed connections logged, would just like to know what's going on

Do my notes in Building Happiness answer that? http://nbviewer.jupyter.org/github/bokeh/bokeh-demos/blob/happiness_v2/happiness/Building%20happiness.ipynb#Things-I-would-improve

> reset seems to weirdly shrink plot size (!)

it's not doing that for me - bug in bokeh? - can you give me browser/bokeh verstion etc.

> does not seems to be auto-ranging when new points are added?

can you clarify? when new points are added I see them appear on other plots if that's what you mean?

> didn't realize date did not get incremented when happiness logged, was confusing, suggest auto-incrementing date

i agree, but is making my fake app user friendly in-scope?

> would be nice if the data was always the "last X days, regardless of when the user runs the demo

yes, suggestions on how to generate fake data in the database gratefully received.

File: happiness/happiness/bokeh_utils.py (hunk 1 above)
Comment (bryevdv) [inline happiness/happiness/bokeh_utils.py]: I meant multiple Bokeh server behind a load balancer. For this approach we'd need some kind of sticky session affinity I think. Just for the future, here is the app I think would be the most valuable to demonstrate:

* Django (whatever) creates a session, transmits credentials or tokens to a shared store to it
* Django gets and uses autoad script tag
* now Django can update the shared store
* bokeh only responds to shared store now (Django never touches bokeh session again)

**Comment by birdsarah** [comment]
> 

>>    didn't realize date did not get incremented when happiness logged, was confusing, suggest auto-incrementing date

> i agree, but is making my fake app user friendly in-scope?

Following-up on my comment about in-scope: I don't think this is a nice demo in the sense of something we want to show off as "look at the capabilities of bokeh etc". I think it's useful code for hooking things together, which is why I'm not too bothered about usability.

**Comment by bryevdv** [comment]
> Do my notes in Building Happiness answer that?

Maybe, hadn't gotten to firing up a notebook yet

> it's not doing that for me - bug in bokeh? - can you give me browser/bokeh verstion etc. 
> can you clarify? when new points are added I see them appear on other plots if that's what you mean?


let me investigate a bit

> i agree, but is making my fake app user friendly in-scope?
 
at least enough so that people don't think it was broken (I thought it was broken until I realized I was logging happiness for the same day over and over)

**Comment by bryevdv** [comment]
> Following-up on my comment about in-scope: I don't think this is a nice demo in the sense of something we want to show off as "look at the capabilities of bokeh etc". I think it's useful code for hooking things together, which is why I'm not too bothered about usability.

Well, like I said, I thought it was broken. Even if it's not super impressive as a thing unto itself, I think we have to clear the "seems to be working" bar.

**Comment by birdsarah** [comment]
Fair enough. I think I just thought of an easy fix too.

File: happiness/happiness/bokeh_utils.py (hunk 1 above)
Comment (birdsarah) [inline happiness/happiness/bokeh_utils.py]: Yes, I have no idea how to do this - I haven't seen any documentation, or heard anything from havoc, about using `pull_session` in a multi-bokeh server setup. That feels like a whole new demo! (or at least my brain hurts from thinking about it :D)

**Comment by birdsarah** [comment]
> reset seems to weirdly shrink plot size (!)

I am seeing this it shrinks or grows plot size depending on whether responsive view is making plot bigger or smaller. This is a bug in bokeh. Reset is resetting the plot_width and plot_height (which I don't think it used to do) and then the responsive resize code is not being called. I've added an issue: https://github.com/bokeh/bokeh/issues/3783

> Well, like I said, I thought it was broken. Even if it's not super impressive as a thing unto itself, I think we have to clear the "seems to be working" bar. 

This is now done, the form now autofills to the date after the last date the user has entered.

**Comment by birdsarah** [comment]
ping @bryevdv, what do you think about merging this?

**Comment by bryevdv** [comment]
OH this fell completely off my radar.. another reason to move it it the main repo. I'd say sure, and then let's plan to move it.

**Pull request merged**